packages/contracts/contracts/voting/token/TokenVoting.sol

```diff
@@ -104,59 +98,68 @@ contract TokenVoting is MajorityVotingBase {
     ) internal override {
         Proposal storage proposal_ = proposals[_proposalId];
 
         // This could re-enter, though we can assume the governance token is not malicious
-        uint256 votingPower = votingToken.getPastVotes(_voter, proposal_.snapshotBlock);
+        uint256 votingPower = votingToken.getPastVotes(
+            _voter,
+            proposal_.configuration.snapshotBlock
+        );
         VoteOption state = proposal_.voters[_voter];
 
         // If voter had previously voted, decrease count
         if (state == VoteOption.Yes) {
-            proposal_.yes = proposal_.yes - votingPower;
+            proposal_.tally.yes = proposal_.tally.yes - votingPower;
         } else if (state == VoteOption.No) {
-            proposal_.no = proposal_.no - votingPower;
+            proposal_.tally.no = proposal_.tally.no - votingPower;
         } else if (state == VoteOption.Abstain) {
-            proposal_.abstain = proposal_.abstain - votingPower;
+            proposal_.tally.abstain = proposal_.tally.abstain - votingPower;
         }
 
         // write the updated/new vote for the voter.
-        if (_choice == VoteOption.Yes) {
-            proposal_.yes = proposal_.yes + votingPower;
-        } else if (_choice == VoteOption.No) {
-            proposal_.no = proposal_.no + votingPower;
-        } else if (_choice == VoteOption.Abstain) {
-            proposal_.abstain = proposal_.abstain + votingPower;
+        if (_voteOption == VoteOption.Yes) {
+            proposal_.tally.yes = proposal_.tally.yes + votingPower;
+        } else if (_voteOption == VoteOption.No) {
+            proposal_.tally.no = proposal_.tally.no + votingPower;
+        } else if (_voteOption == VoteOption.Abstain) {
+            proposal_.tally.abstain = proposal_.tally.abstain + votingPower;
         }
 
-        proposal_.voters[_voter] = _choice;
+        proposal_.voters[_voter] = _voteOption;
 
         emit VoteCast({
             proposalId: _proposalId,
             voter: _voter,
-            choice: uint8(_choice),
+            voteOption: _voteOption,
             votingPower: votingPower
         });
 
-        if (_executesIfDecided && _canExecute(_proposalId)) {
+        if (_tryEarlyExecution && _canExecute(_proposalId)) {
             _execute(_proposalId);
         }
     }
 
     /// @inheritdoc MajorityVotingBase
     function _canVote(uint256 _proposalId, address _voter) internal view override returns (bool) {
         Proposal storage proposal_ = proposals[_proposalId];
+
         return
-            _isVoteOpen(proposal_) && votingToken.getPastVotes(_voter, proposal_.snapshotBlock) > 0;
+            _isVoteOpen(proposal_) &&
+            // Check if voter has voting power.
+            votingToken.getPastVotes(_voter, proposal_.configuration.snapshotBlock) > 0 &&
+            // Check if the vote is a disallowed vote replacement.
+            !(proposal_.configuration.voteMode != VoteMode.VoteReplacement &&
```

Comment:
Same as before, I would spread these heterogeneous conditions into separate `if`+ `return`'s

Reply:
Fixed in 868a1fcb6be7879977da7fe59e724452fb17945d

Reply:
Fixed in 868a1fcb6be7879977da7fe59e724452fb17945d